### ISSUE TITLE
Receive entity: <null> data: <fault> 

### ISSUE THREAD

**Issue by achirkof** [hyperoslo/Sync#261]
Hi @3lvis !
I meet another problem. I receive `entity: <null>` `data: <fault>` when provide data model from one UIViewController to another. I know about Core Data faulting mechanism, but `entity: <null>` looks strange. Looks like I loose a `context`.

Here is my `data model`:
![data model](https://cloud.githubusercontent.com/assets/1292136/17921083/d27daa28-69e0-11e6-9b83-83a08006c4b9.png)

I use a function in first ViewController named `fetchCurrentWorkouts` to retrieve data from Core Data. Here is (some code was skipped):
```Swift
import UIKit
import DATAStack
import CoreData

class WorkoutSectionCell: BaseUserProfileCell, UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {

    // ...
    var workouts: [Workout] = []
    
    override func setupViews() {
        super.setupViews()
        // ...
        fetchCurrentWorkouts()
    }
    
    // retrieve data from Core Data
    func fetchCurrentWorkouts() {
        let request = NSFetchRequest(entityName: "Workout")
        self.workouts = try! DATAStack(modelName: "smartworkout_rest").mainContext.executeFetchRequest(request) as! [Workout]
        self.collectionView.reloadData()
    }
    
    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        let count = self.workouts.count
        return count
    }
    
    // provide data from model to collection view cell 
    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier(workoutCell, forIndexPath: indexPath) as! WorkoutCell
        cell.workout = self.workouts[indexPath.item]
        print(self.workouts[indexPath.item]) // prints result marked "1" on picture
        // here the button which should segue to another controller
        cell.goWorkoutDetailButton.tag = indexPath.item
        cell.goWorkoutDetailButton.addTarget(self, action: #selector(navigateWorkoutDetail), forControlEvents: .TouchUpInside)
        return cell
    }
    
    // here is the problem
    func navigateWorkoutDetail(sender: UIButton) {
        let workout = self.workouts[sender.tag]
        print(workout) // prints result marked "2" on picture
        //self.userProfileVC?.goWorkoutDetail(workout)
    }
    
    // rest collection view related code ...
}
```
Here the interface. Button locate on the same controller.

![interface](https://cloud.githubusercontent.com/assets/1292136/17921583/3ea5b946-69e3-11e6-935d-4ef40d3c2be2.png)

Here the result output:

![result](https://cloud.githubusercontent.com/assets/1292136/17921734/1e4aab2e-69e4-11e6-9a37-0aed54057ab4.png)

As we can see on result line marked "2" linked to first value from the model, because
`smartworkout_rest.Workout: 0x7f8fb9ce71a0>` from first print equal to `<smartworkout_rest.Workout: 0x7f8fb9ce71a0>` from second print. But no entity and no data.

Looks like I forgot something or made a mistake. Could you help me solve this issue?
Thanks.

**Comment by 3lvis** [comment]
Having that aside, maybe Core Data is releasing the object since the controller that invoke it is no longer being used, for example when you push the new controller. I would either just pass the primary key to the new controller and make a new fetch there. Or I would try to use [DATASource](https://github.com/3lvis/DATASource) to see if it helps with the memory related issues.

**Comment by achirkof** [comment]
Oh, my goodness!
Thank you @3lvis for the idea 

> maybe Core Data is releasing the object since the controller that invoke it is no longer being used

The reason was that function `fetchCurrentWorkouts` calls once when view loads and destroy after and remove context. That's why I have `entity: <null>`.
Declaring constant `dataStack` 

```swift
class WorkoutSectionCell: BaseUserProfileCell, UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {

    let dataStack: DATAStack = DATAStack(modelName: "smartworkout_rest")

//... rest code

```

and call it in fetch request

```swift
func fetchCurrentWorkouts() {
        let request = NSFetchRequest(entityName: "Workout")
        self.workouts = try! dataStack.viewContext.executeFetchRequest(request) as! [Workout]
        print(workouts)
        self.collectionView.reloadData()
    }
```
solve the issue.

**Comment by 3lvis** [comment]
It's better if you don't create multiple `DATAStack`s, but reuse the same one. Otherwise unexpected things might happen.

I guess you're using Storyboards, I have a demo on how it's the recommended way of doing it if you're interested.

Project: https://github.com/3lvis/StoryboardDATASourceAndDataStackDemo

**Comment by achirkof** [comment]
I don't use storyboard :-) because it is more easy for me to use code. But anyway it will be very interesting. I'm just studying iOS development and any examples will be very useful for me.
Thanks!

**Comment by 3lvis** [comment]
If you don't use Storyboards then this might be useful

**Comment by achirkof** [comment]
Yes, already studying it. Great example, thanks! You made perfect pod's. I spend several days dealing with native JSON parsing and saving data to Core Data. Sync make this job much simpler and just with several lines of code. Awesome!

**Comment by 3lvis** [comment]
Awww, thanks! I'm glad you like it.
